### PR TITLE
Audio fingerprint protections

### PR DESCRIPTION
**Pull request by jonathanKingston**
**Reviewer:**

## Automated tests:
- [ ] Unit tests
- [ ] Integration tests

###### Reviewer Checklist:
- [ ] **Ensure the PR solves the problem**
- [ ] **Review every line of code**
- [ ] **Ensure the PR does no harm by testing the changes thoroughly**
- [ ] **Get help if you're uncomfortable with any of the above!**
- [ ] Determine if there are any quick wins that improve the implementation

###### PR Author Checklist:
- [ ] Get advice or leverage existing code
- [ ] Agree on technical approach with reviewer (if the changes are nuanced)
- [ ] Ensure that there is a testing strategy (and documented non-automated tests)
- [ ] Ensure there is a documented monitoring strategy (if necessary)
- [ ] Consider systems implications

### CODE REVIEW

File: shared/js/background/utils.es6.js [hunk 1]
```diff
@@ -154,13 +154,29 @@ function isBroken (url) {
...
+            brokenFeatures.push(feature)
+        }
+        if (isBrokenList(url, tdsStorage.fingerprinting[feature].sites || [])) {
+            brokenFeatures.push(feature)
+        }
+    }
+    return brokenFeatures
 }
 
+/*
```
Comment (Charlie-belmer) [inline shared/js/background/utils.es6.js:171]: We should probably remove this before release if it's not needed.

**Review by Charlie-belmer — COMMENTED** [review]
Looking good so far - I see some commented code that look like they aren't needed anymore? Other than that, it's looking pretty good from a code perspective - haven't tried it yet.

File: shared/js/content-scope/fingerprint.js [hunk 2]
```diff
@@ -145,3 +145,93 @@ function initCanvasProtection (args) {
...
+    function transformArrayData (channelData) {
+        const cdSum = channelData.reduce((sum, v) => {
+            return sum + v
+        }, 0)
+        const audioKey = getDataKeySync(sessionKey, domainKey, cdSum)
+        let pixel = audioKey.charCodeAt(0)
+        for (const i in audioKey) {
+            let byte = audioKey.charCodeAt(i)
+            for (let j = 8; j >= 0; j--) {
+                const pixelCanvasIndex = pixel % channelData.length
```
Comment (Charlie-belmer) [inline shared/js/content-scope/fingerprint.js:165]: Is this loop close enough to the canvas algo that we could combine into a function? We should probably rename this var to audio if you keep it as is..

File: shared/js/content-scope/fingerprint.js [hunk 3]
```diff
@@ -145,3 +153,84 @@ function initCanvasProtection (args) {
...
+            // Anything we do here should be caught and ignored silently
+            try {
+                transformArrayData(channelData, domainKey, sessionKey)
+            } catch {
+            }
+            return channelData
+        }
+    })
+
+    const canvasMethods = ['getByteTimeDomainData', 'getFloatTimeDomainData', 'getByteFrequencyData', 'getFloatFrequencyData']
```
Comment (jonathanKingston) [inline shared/js/content-scope/fingerprint.js:220]: Note to self, rename this.

**Review by Charlie-belmer — APPROVED** [review]
Looks good to me here. I took it for a spin and audio seems to be working well while blocking fingerprinting. Nice work! Merge when ready.

**Pull request merged**